### PR TITLE
Adds JSON Patch (RFC 6902) support

### PR DESCRIPTION
**Pull request by biofractal**
PATCH support has been added via the [mongoose-json-patch](https://github.com/winduptoy/mongoose-json-patch) mongoose plugin.

It appears that plugins cannot be attached after the mongoose model has been generated (?) Therefore, instead of sending in a model, the schema must be supplied instead. The model patch middleware is then attached and the model subsequently generated.

### CODE REVIEW

File: Resource.js [hunk 1]
```diff
@@ -341,6 +349,28 @@ module.exports = function(app, route, modelName, model) {
...
 
     /**
+     * Patch (Partial Update) a resource.
+     */
+    patch: function(options) {
+      this.methods.push('patch');
+      app.patch.apply(app, this.register(this.route + '/:' + this.name + 'Id', function(req, res, next) {
+        var query = req.modelQuery || this.model;
+        query.findOne({"_id": req.params[this.name + 'Id']}, function(err, item) {
+          if (err) return this.respond(res, 500, err);
```
Comment (travist) [inline Resource.js]: This should say ```return this.setResponse(res, {status: 500, error: err}, next);```

File: Resource.js [hunk 2]
```diff
@@ -341,6 +349,28 @@ module.exports = function(app, route, modelName, model) {
...
+          if (!item) return this.respond(res, 404);
```
Comment (travist) [inline Resource.js]: This should say ```return this.setResponse(res, {status: 404}, next);```

**Comment by travist** [comment]
Looks good!  Just make those few changes and I will pull this in.

Thanks for the contribution!

**Comment by biofractal** [comment]
OK, done. I noticed the error call below was in the old style as well so I fixed that up too.

File: Resource.js [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
+var patcher = require('mongoose-json-patch');
 
-module.exports = function(app, route, modelName, model) {
+module.exports = function(app, route, modelName, schema) {
+
+  // Attach the mongoose-json-patch plugin
+  schema.plugin(patcher);
+
+  // Generate the mongoose model
+  var model= mongoose.model(modelName, schema);
```
Comment (travist) [inline Resource.js]: I have an opinion about this... 

The reason why I was keeping the model declaration out of this library is because it is common for someone to declare the model somewhere else in their library where it is used in other placed.  Here we would be creating a separate instance of that model which make consume more memory as well as cause unwanted consequences...

Were you having problems passing the ```mongoose.model``` class to this library?

**Comment by biofractal** [comment]
I agree that the model should be passed in rather than the schema. I had no problem with this until I tried to attach the patch middleware. I found that I could not attach the patch middleware to the schema of an existing model (via the `model.schema` property at least). This is why I ended up passing in the schema rather than the model.

Attaching the middleware to the schema then generating the model does work so I stopped there. I did not investigate fully whether it is actually possible to attach middleware to the schema of a pre-existing model. If this is possible then we can revert to passing in the model.

**Comment by travist** [comment]
That makes sense... I am curious if you can get access to this schema from the model instance? I would have to investigate as well. My only concern is that this is a "breaking" change and we would need to increase the major version to take it. I want patch capabilities, but would like to look for alternatives than pushing out a "breaking" change.

**Comment by biofractal** [comment]
I looked into it a bit more and found out two things:

1. It is easy to do the patching ourselves using the [fast-json-patch](https://github.com/Starcounter-Jack/JSON-Patch) package. In this way we side-step the whole schema middleware issue and restore your original model-centric design - good news. These changes have now been made and pushed onto the `patch` branch if you want to review. It is simpler and appears to be working as before.

2. There is a problem with `fast-json-patch` (see issue below). It is not currently working to the [RFC-6902](https://tools.ietf.org/html/rfc6902) spec as it does not appear to be respecting failing `test` ops. This is a big pity because `test` ops are critical to any reasonable patch strategy (imo). The package has a lot traction and is performant so I would like to stick with it though. Perhaps you could add your weight to calls for the damn thing to be fixed (or check out the code to see if I have just not understood how to use it).
   * Issue: [apply() only returns result of last operation](https://github.com/Starcounter-Jack/JSON-Patch/issues/64)

**Comment by biofractal** [comment]
The new code first run all the test ops in the patch sequence before attempting to apply any changes. If any test op fails then no changes are applied to the document. Hopefully you will be able to remove this when fast-json-patch finally gets fixed.

File: npm-debug.log [hunk 4]
```diff
@@ -0,0 +1,72 @@
+0 info it worked if it ends with ok
```
Comment (travist) [inline npm-debug.log]: Please do not commit ```npm-debug.log```.

**Comment by travist** [comment]
This looks good. Thanks again for your contribution.

Just a few things before I pull this in... 1.) I would like a test (in the test folder) that tests for this new capability.  2.) Please remove the npm-debug.log from your commit.

Once we get those in place, I will happily accept this pull request.

Thanks again!

Travis.

**Comment by biofractal** [comment]
I have removed the log file and added the unit-tests as requested.

File: Resource.js [hunk 5]
```diff
@@ -80,10 +82,13 @@ module.exports = function(app, route, modelName, model) {
                 return _.pick(error, 'path', 'name', 'message');
               })
             });
-          case 404:
```
Comment (travist) [inline Resource.js]: Why did you remove the `404` response? This will break all the 404 responses for content not found.

**Comment by biofractal** [comment]
Its fixed now. That code was part of an experiment I did some time ago, I have no idea how it re-appeared. Some unexplained weirdness with my Git branch. I just lost all the new patch code as well and had to restore that from history. Maybe sublime didn't update when I swapped branches and then I saved some chimeric version - or something. Anyway, it should be fine now. Sorry for the mess up.

**Comment by travist** [comment]
Looks great!  Thank you for your contribution!
